### DISCUSSION TITLE
[next-auth][error][JWT_SESSION_ERROR], how to fix this error in next-auth

### DISCUSSION THREAD

**Discussion by IsharaWeerasinghege** [nextauthjs/next-auth#8957] [Help]
### Summary

i am using next-auth for myy nextjs web application. in local development my application are working fine but the bellow error thrown in production build. how to fix this issues. i will try different methods for fixing that issues, but still cant fixed this issue.

`
[next-auth][error][JWT_SESSION_ERROR] 
https://next-auth.js.org/errors#jwt_session_error Operation `users.findOne()` buffering timed out after 10000ms {
  message: 'Operation `users.findOne()` buffering timed out after 10000ms',
  stack: 'MongooseError: Operation `users.findOne()` buffering timed out after 10000ms\n' +
    '    at Timeout.<anonymous> (G:\\Next project\\videoclay\\node_modules\\mongoose\\lib\\drivers\\node-mongodb-native\\collection.js:186:23)\n' +
    '    at listOnTimeout (node:internal/timers:564:17)\n' +
    '    at process.processTimers (node:internal/timers:507:7)',
  name: 'MongooseError'
}
`

mongo database connection.
db.js

`
import mongoose from 'mongoose';

const connection = {};
mongoose.set('strictQuery', true);

async function connectDb() {
    if (connection.isConnected){
        console.log('Already connected to the database');
        return;
    }

    if (mongoose.connections.length > 0) {
        connection.isConnected = mongoose.connections[0].readyState;
        if (connection.isConnected === 1) {
            console.log('Use existing connection');
            return;
        }
        await mongoose.disconnect();
    }

    const db = await mongoose.connect(process.env.MONGODB_URL, {
        useNewUrlParser: true,
        useUnifiedTopology: true,
    })
    console.log('New connection to the database');
    connection.isConnected = db.connections[0].readyState;
}

async function disconnectDb() {
    if (connection.isConnected) {
        if (process.env.NODE_ENV === 'production') {
            await mongoose.disconnect();
            connection.isConnected = false;
        } else {
            console.log('Not disconnected');
        }
    }
}

const db = { connectDb, disconnectDb };

export default db;
`

mongo database connation for next-auth
mongodb.js

`
import {MongoClient} from "mongodb";

const uri = process.env.MONGODB_URL;
const options = {
    useUnifiedTopology: true,
    // useNewUrlParser: true,
};

let client;
let clientPromise;

if (!process.env.MONGODB_URL) {
    throw new Error("Please add your Mongo URI to .env.local");
}

if (process.env.NODE_ENV === "development") {
    if (!global._mongoClientPromise) {
        client = new MongoClient(uri, options);
        global._mongoClientPromise = client.connect();
    }
    clientPromise = global._mongoClientPromise;
} else {
    client = new MongoClient(uri, options);
    clientPromise = client.connect();
}

export default clientPromise;
`

[next-auth].js

`
import NextAuth from "next-auth";
import FacebookProvider from "next-auth/providers/facebook";
import GoogleProvider from "next-auth/providers/google";
import CredentialsProvider from "next-auth/providers/credentials";
import User from "../../../models/User";
import clientPromise from "./lib/mongodb";
import bcrypt from "bcrypt";
import db from "../../../utils/db";
import {MongoDBAdapter} from "@auth/mongodb-adapter";

export default NextAuth({
    adapter: MongoDBAdapter(clientPromise),
    providers: [
        CredentialsProvider({
            name: "Credentials",
            credentials: {
                username: { label: "Username", type: "text", placeholder: "jsmith" },
                password: { label: "Password", type: "password" },
            },
            async authorize(credentials, req) {
                const email = credentials.email;
                const password = credentials.password;
                await db.connectDb();
                const user = await User.findOne({ email });
                if (user) {
                    return SignInUser({ password, user });
                } else {
                    throw new Error("This email does not exist.");
                }
            },
        }),
        FacebookProvider({
            clientId: process.env.FACEBOOK_ID,
            clientSecret: process.env.FACEBOOK_SECRET,
        }),
        GoogleProvider({
            clientId: process.env.GOOGLE_ID,
            clientSecret: process.env.GOOGLE_SECRET,
        }),
    ],
    callbacks: {
        async session({ session, token }) {
            const user = await User.findById(token.sub);
            session.user.id = token.sub || user._id.toSting();
            session.user.role = user.role || "user";
            token.role = user.role || "user";

            return session;
        },
    },
    pages: {
        signIn: "/signin",
    },
    session: {
        strategy: "jwt",
    },
    secret: process.env.JWT_SECRET,
});

const SignInUser = async ({ password, user }) => {
    if (!user.password) {
        throw new Error("Please enter your password.");
    }
    const testPassword = await bcrypt.compare(password, user.password);
    if (!testPassword) {
        throw new Error("Email or password is wrong!");
    }
    return user;
};
`

**Comment by Fox333-Lab** [comment]
Facing the same exact issue, it was working fine a day back and today its throwing error.
Any suggestions, please

**Comment by sem22-dev** [comment]
Got the same error.

This happens when your database is not connected, 

In my case , it got fixed when i  add "await connectMongoDB();" just before calling "await User.findOne()"

example : 
//connectMongoDB looks like this : 

    import mongoose from "mongoose";
    export const connectMongoDB = async () => {
      try {
        await mongoose.connect(process.env.MONGODB_URI);
        console.log("Connected to MongoDB");
      } catch (error) {
        console.log("Error connecting to MongoDB: ", error);
      }
    };

//part of callback function in auth.ts

      import { connectMongoDB } from '@/lib/mongodb';
      async session({ session }) {
      await connectMongoDB();
        if (session?.user?.email) {
          const sessionUser = await User.findOne({ email: session.user.email });
          if (sessionUser) {
            // Safely assign id only if sessionUser is found
            session.user.id = sessionUser._id.toString();
            console.log('userId', session.user.id);
          }
        }
        return session;
      },

> Facing the same exact issue, it was working fine a day back and today its throwing error. Any suggestions, please

**Reply by tekusab-master** [reply]
Thank you very much for this! This definitely fixed my problem! Happy coding bro!
